cfgov/unprocessed/apps/youth-employment-success/js/util.js

```diff
@@ -0,0 +1,56 @@
+let UNDEFINED;
+
+const REDUCER_RETURN_ERROR = 'Reducer must return a state object';
+const processStateFromReducers = reducers => (
+  state = UNDEFINED,
+  action = { type: null }
+) => {
+  /**
+   * If reducers is not an object
+   */
+  const reducersKeys = Object.keys( reducers );
+  const finalState = {};
+  let hasChanged = false;
+
+  for ( let i = 0; i < reducersKeys.length; i++ ) {
+    const key = reducersKeys[i];
+    const reducer = reducers[key];
+
+    /**
+     * if we didnt already get a composite state object, e.g. if this is
+     * the first time the reducers are running, default to undefined
+     * and force the reducer to use it's default state
+     */
+    const previousState = state ? state[key] : state;
+    let argumentsToApply = [ action ];
+    argumentsToApply = [ previousState, ...argumentsToApply ];
+
+    const nextState = reducer.apply( reducer, argumentsToApply );
+
+    if ( typeof nextState === 'undefined' ) {
+      throw new Error( REDUCER_RETURN_ERROR );
+    }
+
+    finalState[key] = nextState;
+    hasChanged = hasChanged || nextState !== previousState;
+  }
+
+  return hasChanged ? finalState : state;
+};
+
+export const combineReducers = ( reducers = {} ) => {
```

Comment:
Lot of nested logic here. Okay, but would be clearer if the complexity was reduced. Food for thought.